### PR TITLE
Document vector class

### PR DESCRIPTION
**Pull request by quantumsteve**
Please review the [developer documentation](https://github.com/project-asgard/asgard/wiki/developing)
on the wiki of this project that contains help and requirements.

## Proposed changes

Describe what this PR changes and why.  If it closes an issue, link to it here
with [a supported keyword](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

Document members of the vector class.

## What type(s) of changes does this code introduce?
_Put an `x` in the boxes that apply._

- [ ] Bugfix
- [ ] New feature
- [ ] Code style update (formatting, renaming)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [x] Documentation content changes
- [ ] Other (please describe):

### Does this introduce a breaking change?

- [ ] Yes
- [x] No

## Checklist

_Put an x in the boxes that apply. You can also fill these out after creating
the PR. If you're unsure about any of them, don't hesitate to ask.  This is
simply a reminder of what we are going to look for before merging your code._

- [x] this PR is up to date with current the current state of 'develop'
- [x] code added or changed in the PR has been clang-formatted
- [ ] this PR adds tests to cover any new code, or to catch a bug that is being fixed
- [x] documentation has been added (if appropriate)

### CODE REVIEW

File: src/tensors.hpp [hunk 1]
```diff
@@ -241,105 +319,198 @@ class vector
...
   // math operators
-  //
+
+  /*! element-wise addition
+   *  \param right elements on rhs for addition.
+   *  \return vector with results of element-wise addition.
+   */
   template<mem_type omem, resource r_ = resrc, typename = enable_for_host<r_>>
   vector<P> operator+(vector<P, omem> const &right) const;
+  /*! element-wise subtraction
```
Comment (mkstoyanov) [inline src/tensors.hpp:353]: Silly question, are we using this and why are we not using BLAS for these kind of operations.

File: src/tensors.hpp [hunk 2]
```diff
@@ -241,105 +319,198 @@ class vector
...
   template<mem_type omem, resource r_ = resrc, typename = enable_for_host<r_>>
   vector<P> operator-(vector<P, omem> const &right) const;
+  /*! dot product of two vectors
+   *  \param right elements on rhs for dot product.
+   *  \return scalar result of dot product.
+   */
   template<mem_type omem, resource r_ = resrc, typename = enable_for_host<r_>>
   P operator*(vector<P, omem> const &) const;
-
+  /*! vector*matrix product
```
Comment (mkstoyanov) [inline src/tensors.hpp:365]: This is a bad idea, it will force a new allocation, we don't want to have a new allocation every time we do a matrix-vector product.

It's good to see the documentation, we can now see these issues.

File: src/tensors.hpp [hunk 3]
```diff
@@ -241,105 +319,198 @@ class vector
...
+   *  \return pointer to private data
+   */
   P *data(int const elem = 0) const { return data_ + elem; }
 
-  // this is to allow specific other types to access the private ref counter of
-  // owners - specifically, we want to allow a matrix<view> to be made from a
-  // vector<owner/view>
+  /*! this is to allow specific other types to access the private ref counter of
+   * owners - specifically, we want to allow a matrix<view> to be made from a
+   * vector<owner/view> \param access badge limiting public access. \return
```
Comment (mkstoyanov) [inline src/tensors.hpp]: `\param` and `\return` should be on separate lines.

File: src/tensors.hpp [hunk 4]
```diff
@@ -241,105 +319,198 @@ class vector
...
-  get_ref_count(access_badge<matrix<P, mem_type::view, resrc>> const)
+  get_ref_count(access_badge<matrix<P, mem_type::view, resrc>> const access)
   {
+    ignore(access);
     return ref_count_;
   }
-
+  /*! this is to allow specific other types to access the private ref counter of
+   * owners - specifically, we want to allow a matrix<view> to be made from a
+   * vector<owner/view> \param access badge limiting public access. \return
```
Comment (mkstoyanov) [inline src/tensors.hpp]: New line for `param` and `return`

File: src/tensors.hpp [hunk 5]
```diff
@@ -241,105 +319,198 @@ class vector
...
   template<resource r_ = resrc, typename = enable_for_host<r_>>
-  void print(std::string const label = "") const;
+  void print(std::string_view const label = "") const;
 
+  /*! Dumps to file a vector that can be read straight into octave
+   * \param filename name
+   */
   template<resource r_ = resrc, typename = enable_for_host<r_>>
-  void dump_to_octave(char const *) const;
+  void dump_to_octave(std::filesystem::path const &filename) const;
```
Comment (mkstoyanov) [inline src/tensors.hpp:445]: Why change to `path` and why in the documentation PR?

File: src/tensors.hpp [hunk 6]
```diff
@@ -241,105 +319,198 @@ class vector
...
+   * \param index first element to set
+   * \param sub_vector container of elements to assign
+   * \return reference to this
+   */
   template<mem_type omem, mem_type m_ = mem,
            typename = disable_for_const_view<m_>, resource r_ = resrc,
            typename = enable_for_host<r_>>
-  vector<P, mem> &set_subvector(int const, vector<P, omem> const);
+  vector<P, mem> &
+  set_subvector(int const index, vector<P, omem> const sub_vector);
```
Comment (mkstoyanov) [inline src/tensors.hpp:463]: `sub_vector` should be captured by ref, not copied. We should fix this in another PR.

File: src/tensors.hpp [hunk 7]
```diff
@@ -577,7 +751,7 @@ class matrix
   void print(std::string const label = "") const;
 
   template<resource r_ = resrc, typename = enable_for_host<r_>>
-  void dump_to_octave(char const *name) const;
+  void dump_to_octave(std::filesystem::path const &filename) const;
```
Comment (mkstoyanov) [inline src/tensors.hpp:756]: Same question about the change in path.

File: src/tensors_tests.cpp [hunk 8]
```diff
@@ -849,9 +849,9 @@ TEMPLATE_TEST_CASE("fk::vector utilities", "[tensors]", double, float, int)
         (std::istreambuf_iterator<char>(data_stream_cv)),
         std::istreambuf_iterator<char>());
 
-    std::remove("test_out.dat");
-    std::remove("test_out_v.dat");
-    std::remove("test_out_cv.dat");
+    REQUIRE(std::filesystem::remove("test_out.dat"));
```
Comment (mkstoyanov) [inline src/tensors_tests.cpp:852]: Is this why we changed to a `std::filesystem`? If so, it probably belongs to another PR.

**Review by mkstoyanov — CHANGES_REQUESTED** [review]
Mostly minor typos.

File: src/tensors.hpp (hunk 1 above)
Comment (quantumsteve) [reply src/tensors.hpp:353]: Calls to BLAS, LAPACK and ScaLAPACK are in the [lib_dispatch namespace](https://github.com/project-asgard/asgard/blob/develop/src/lib_dispatch.hpp). I think tensors calls level 2 and level 3 BLAS functions like gemv and gemm.

File: src/tensors.hpp (hunk 2 above)
Comment (quantumsteve) [reply src/tensors.hpp:365]: Yes, some of these member functions are easy to use incorrectly.

Functions in [fast_math](https://github.com/project-asgard/asgard/blob/4169a383f81945e8fbe99747489dfe574a6cc0d1/src/fast_math.hpp#L131) are more explicit.

File: src/tensors.hpp (hunk 3 above)
Comment (quantumsteve) [reply src/tensors.hpp]: I'll look into our clang-format settings and see if I can keep this from happening.

File: src/tensors.hpp (hunk 5 above)
Comment (quantumsteve) [reply src/tensors.hpp:445]: A more specific type is a form of documentation. 

We moved filenames from `std::string` and `const char*` to `std::fileystem::path` in #423 and this is one I appear to have missed. Do you need this in a separate PR?

File: src/tensors_tests.cpp (hunk 8 above)
Comment (quantumsteve) [reply src/tensors_tests.cpp:852]: We moved filenames from std::string and const char* to std::fileystem::path in https://github.com/project-asgard/asgard/pull/423.

I saw an opportunity to add another check. The previous code should have checked that `std::remove` returns `0`.

Comment (mkstoyanov) [reply src/tensors_tests.cpp:852]: Don't bother changing this now, just for the future, it's better to have many small PRs than large blobs with all kinds of changes.

File: src/tensors.hpp (hunk 5 above)
Comment (mkstoyanov) [reply src/tensors.hpp:445]: Should have been separate, but don't bother now.

File: src/tensors.hpp (hunk 3 above)
Comment (mkstoyanov) [reply src/tensors.hpp]: That's odd, why are they reformatting the comments?

File: src/tensors.hpp (hunk 2 above)
Comment (mkstoyanov) [reply src/tensors.hpp:365]: I'll make a note to discuss this on the next meeting. There's little point to keep inefficient methods in the code, have to check if those are actually used anywhere.

**Comment by mkstoyanov** [comment]
@quantumsteve I guess the only real issue here is the clang-format one, the rest is more of comments and points to discuss in the future.

File: src/tensors.hpp (hunk 3 above)
Comment (quantumsteve) [reply src/tensors.hpp]: https://github.com/llvm/llvm-project/issues/43831

File: .clang-format [hunk 9]
```diff
@@ -107,5 +107,6 @@ SpacesInSquareBrackets: false
 Standard:        Cpp11
 TabWidth:        8
 UseTab:          Never
+# Avoid formatting (breaking) doxygen
+CommentPragmas:  '^\\.+'
```
Comment (mkstoyanov) [inline .clang-format:111]: The whole idea behind formatting should be taken as a suggestion, enforcing it with an automated tool is just too much.

@quantumsteve  good job on catching this one.

**Review by mkstoyanov — APPROVED** [review]
Let's just wait for the CI
